### PR TITLE
Move configuration of children to __post_configure__

### PR DESCRIPTION
**Pull request by jneeven**

### CODE REVIEW

File: zookeeper/core/component.py [hunk 1]
```diff
@@ -1,5 +1,4 @@
-"""
-Components are generic, modular classes designed to be easily configurable.
+"""Components are generic, modular classes designed to be easily configurable.
```
Comment (AdamHillier) [inline zookeeper/core/component.py:1]: I don't mind changing the docstring format, but can we do it in a different PR and apply it to the whole repo? It will make review of this one easier.

File: zookeeper/core/component.py [hunk 2]
```diff
@@ -239,7 +237,10 @@ def _wrap_setattr(component_cls: Type) -> None:
     def wrapped_fn(instance, name, value):
         try:
             value_defined_on_class = getattr(component_cls, name)
-            if isinstance(value_defined_on_class, Field):
+            if (
+                isinstance(value_defined_on_class, Field)
+                and value_defined_on_class.__component_configured__
```
Comment (AdamHillier) [inline zookeeper/core/component.py:242]: `.__component_configured__` won't exist here because `value_defined_on_class` is a `Field` instance, that is defined on the class object, not the instance object. I think might be where the attribute error that's being thrown in the tests is coming from.

Is what you intend to do here to check if the current component is configured or not? I don't think it makes sense to allow setting `parent.child` if `parent` has been configured but `child` has not.

Is this functionality required for this PR, because I worry that it's going to introduce bugs. The field access logic relies on the fact that the instance dict of a component doesn't contain any keys that correspond to field names.

File: zookeeper/core/component.py [hunk 3]
```diff
@@ -368,6 +367,44 @@ def __component_init__(instance, **kwargs):
...
+    for field_name, field_config in conf.items():
+        configure(getattr(instance, field_name), field_config)
+
+
+def __default_post_configure__(instance, conf):
+    print(conf)
+    instance.__configure_children__(conf)
+
+
+def _wrap_post_configure(component_cls: Type) -> None:
```
Comment (AdamHillier) [inline zookeeper/core/component.py:385]: Although the `-> None` return type annotation is correct, could we remove it here to be consistent with the other functions in this file?

File: zookeeper/core/component.py (hunk 2 above)
Comment (jneeven) [reply zookeeper/core/component.py:242]: Ah yes, this should be `instance.__component_configured__`. It is indeed pretty necessary for this PR to enable doing anything in a user-friendly way...

> The field access logic relies on the fact that the instance dict of a component doesn't contain any keys that correspond to field names.

What do you mean by this?

Comment (AdamHillier) [reply zookeeper/core/component.py:242]: https://github.com/larq/zookeeper/blob/b79baea374856facf5e43cf000dc3f70df4f075c/zookeeper/core/component.py#L137-L143

If a value exists in the instance dict corresponding to the name of a field then the `getattribute` function will return it directly. I suspect this means the CLI overriding would break if you did `setattr` of a field value, even if it's done prior to configuration.

Comment (jneeven) [reply zookeeper/core/component.py:242]: It works for me, but let's discuss in detail this afternoon :+1:

Comment (AdamHillier) [reply zookeeper/core/component.py:242]: If this behaviour is cruicial, then I think what you need to do is insert the value into the dict `__component_instantiated_field_values__`.

However, I do think this behaviour is separate from the post-configure stuff that you are doing in this PR, so I would rather break it into another PR. They are two different features, even if they might be used together sometimes. In a different PR we can give proper attention to getting this right, and adding appropriate test cases.

**Comment by jneeven** [comment]
Closing this in favor of a much cleaner solution; will make a PR soon
